File: turms-client-swift/Sources/TurmsClient/Driver/TurmsDriver.swift
import Foundation
import PromiseKit
import Starscream

public class TurmsDriver {
    private static let REQUEST_ID_FIELD = "rid"
    private static let USER_ID_FIELD = "uid"
    private static let PASSWORD_FIELD = "pwd"
    private static let DEVICE_TYPE_FIELD = "dt"
    private static let USER_ONLINE_STATUS_FIELD = "us"
    private static let USER_LOCATION_FIELD = "loc"
    private static let USER_DEVICE_DETAILS = "dd"
    private static let LOCATION_SPLIT = ":"

    private var heartbeatInterval: Int = 20

    private var websocket: WebSocket?
    private var heartbeatTimer: Timer?
    private var isConnected = false
    
    public var onMessage: ((TurmsNotification) -> Void)?
    public var onClose: ((Bool, TurmsStatusCode, Error?, UInt16?) -> Void)?
    private var onDisconnectResolver: Resolver<Void>?

    private var url: String
    private var connectionTimeout: Int
    private var minRequestsInterval: Int?
    private var requestsMap: [Int64: Resolver<TurmsNotification>] = [:]
    private var lastRequestDate = Date(timeIntervalSince1970: 0)
    private var isLastRequestHeartbeat = false
    private var userId: Int64?
    private var password: String?
    
    private var address: String?
    private var sessionId: String?

    public init(_ url: String? = nil, connectionTimeout: Int? = nil, minRequestsInterval: Int? = nil) {
        self.url = url ?? "ws://localhost:9510"
        self.connectionTimeout = connectionTimeout ?? 10
        self.minRequestsInterval = minRequestsInterval
    }

    public func sendHeartbeat() -> Promise<Void> {
        return Promise { seal in
            if connected() {
                setLastRequestRecord(true, Date())
                websocket!.write(data: Data()) { seal.fulfill(()) }
            } else {
                seal.reject(TurmsBusinessError(.clientSessionAlreadyEstablished))
            }
        }
    }

    public func connected() -> Bool {
        return websocket != nil && isConnected
    }

    public func disconnect() -> Promise<Void> {
        return Promise { seal in
            if connected() {
                onDisconnectResolver = seal
                websocket!.disconnect()
            } else {
                seal.reject(TurmsBusinessError(.clientSessionAlreadyEstablished))
            }
        }
    }

    public func connect(userId: Int64, password: String, url: String? = nil, connectionTimeout: Int = 10) -> Promise<Void> {
        return Promise { seal in
            if connected() {
                seal.reject(TurmsBusinessError(.clientSessionAlreadyEstablished))
            } else {
                self.userId = userId
                self.password = password
                var request = URLRequest(
                    url: URL(string: url ?? self.url)!,
                    timeoutInterval: TimeInterval(connectionTimeout))
                request.addValue(String(userId), forHTTPHeaderField: TurmsDriver.USER_ID_FIELD)
                request.addValue(password, forHTTPHeaderField: TurmsDriver.PASSWORD_FIELD)
                websocket = WebSocket(request: request)
                websocket!.onEvent = { event in
                    switch event {
                        case .binary(let data):
                            let notification = try! TurmsNotification(serializedData: data)
                            if notification.hasData, case .session = notification.data.kind! {
                                self.address = notification.data.session.address
                                self.sessionId = notification.data.session.sessionID
                            } else if notification.hasRequestID {
                                let requestId = notification.requestID.value
                                let handler = self.requestsMap[requestId]
                                if notification.hasCode {
                                    let code = notification.code.value
                                    if TurmsStatusCode.isSuccess(code) {
                                        handler?.fulfill(notification)
                                    } else {
                                        handler?.reject(TurmsBusinessError(code))
                                    }
                                } else {
                                    handler?.fulfill(notification)
                                }
                            }
                            if self.onMessage != nil {
                                self.onMessage!(notification)
                            }
                        case .connected:
                            self.onWebsocketOpen()
                            seal.fulfill(())
                        case .disconnected(let reason, let websocketCode):
                            self.onWebsocketClose(Int(reason), websocketCode)
                            self.onDisconnectResolver?.fulfill(())
                        case .cancelled:
                            self.onWebsocketClose()
                            self.onDisconnectResolver?.fulfill(())
                        case .error(let error):
                            self.onWebsocketError(error as! HTTPUpgradeError)
                        default: break
                    }
                }
                websocket!.connect()
            }
        }
    }

    public func resetHeartBeatTimer() {
        heartbeatTimer?.invalidate()
        heartbeatTimer = Timer.scheduledTimer(
            timeInterval: TimeInterval(heartbeatInterval),
            target: self,
            selector: #selector(checkAndSendHeartbeat),
            userInfo: nil,
            repeats: true)
    }

    @objc func checkAndSendHeartbeat() {
        let difference = Date().timeIntervalSince1970 - lastRequestDate.timeIntervalSince1970
        if minRequestsInterval == nil || Int(round(difference)) > minRequestsInterval! {
            sendHeartbeat()
        }
    }

    func send(_ populator: (inout RequestBuilder) -> ()) -> Promise<TurmsNotification> {
        var builder = RequestBuilder()
        populator(&builder)
        let request = builder
            .id(generateRandomId())
            .build()
        return send(request)
    }

    func send(_ request: TurmsRequest) -> Promise<TurmsNotification> {
        return Promise { seal in
            if connected() {
                let now = Date()
                if minRequestsInterval == nil || Int(round(now.timeIntervalSince1970 - lastRequestDate.timeIntervalSince1970)) > minRequestsInterval! {
                    setLastRequestRecord(false, now)
                    let data = try! request.serializedData()
                    resetHeartBeatTimer()
                    requestsMap.updateValue(seal, forKey: request.requestID.value)
                    websocket?.write(data: data)
                } else {
                    seal.reject(TurmsBusinessError(.clientRequestsTooFrequent))
                }
            } else {
                seal.reject(TurmsBusinessError(.clientSessionHasBeenClosed))
            }
        }
    }

    private func setLastRequestRecord(_ isLastRequestHeartbeat: Bool, _ lastRequestDate: Date) {
        self.isLastRequestHeartbeat = isLastRequestHeartbeat
        self.lastRequestDate = lastRequestDate
    }

    private func generateRandomId() -> Int64 {
        var id: Int64
        repeat {
            id = Int64.random(in: 0...Int64.max)
        } while requestsMap.keys.contains(id)
        return id
    }

    private func onWebsocketOpen() {
        resetHeartBeatTimer()
        isConnected = true
    }

    private func onWebsocketClose(_ turmsStatusCode: Int? = nil, _ webSocketCode: UInt16 = 1000) {
        let wasLogged = isConnected
        isConnected = false
        heartbeatTimer?.invalidate()
        onClose?(wasLogged, TurmsStatusCode(rawValue: turmsStatusCode ?? -1)!, nil, webSocketCode)
    }

    private func onWebsocketError(_ error: HTTPUpgradeError) {
        let wasLogged = isConnected
        isConnected = false
        heartbeatTimer?.invalidate()
        if !wasLogged, userId != nil, password != nil {
            switch error {
                case .notAnUpgrade(let code):
                    if code == 307 {
                        connect(userId: userId!, password: password!, url: url)
                    }
                    fallthrough
                default:
                    onClose?(wasLogged, .failed, error, nil)
            }
            connect(userId: userId!, password: password!, url: url)
        } else {
            onClose?(wasLogged, .failed, error, nil)
        }
    }
}

File: turms-client-java/src/main/java/im/turms/client/driver/TurmsDriver.java
package im.turms.client.driver;

import com.google.protobuf.Descriptors;
import com.google.protobuf.Int64Value;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Message;
import im.turms.client.common.Function4;
import im.turms.client.common.StringUtil;
import im.turms.client.common.TurmsLogger;
import im.turms.client.util.ProtoUtil;
import im.turms.common.TurmsCloseStatus;
import im.turms.common.TurmsStatusCode;
import im.turms.common.constant.DeviceType;
import im.turms.common.constant.UserStatus;
import im.turms.common.exception.TurmsBusinessException;
import im.turms.common.model.bo.user.UserLocation;
import im.turms.common.model.dto.notification.TurmsNotification;
import im.turms.common.model.dto.request.TurmsRequest;

import javax.annotation.Nullable;
import javax.validation.constraints.NotNull;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.net.http.WebSocketHandshakeException;
import java.nio.ByteBuffer;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.Function;
import java.util.logging.Level;

import static java.util.AbstractMap.SimpleEntry;

public class TurmsDriver {
    private static final Integer HEARTBEAT_INTERVAL = 20 * 1000;
    public static final String REQUEST_ID_FIELD = "rid";
    public static final String USER_ID_FIELD = "uid";
    public static final String PASSWORD_FIELD = "pwd";
    public static final String DEVICE_TYPE_FIELD = "dt";
    public static final String USER_ONLINE_STATUS_FIELD = "us";
    public static final String USER_LOCATION_FIELD = "loc";
    public static final String USER_DEVICE_DETAILS = "dd";
    private static final String LOCATION_SPLIT = ":";

    private Integer heartbeatInterval;

    private WebSocket websocket;
    private boolean isSessionEstablished;
    private final ScheduledExecutorService heartbeatTimer = Executors.newSingleThreadScheduledExecutor();
    private ScheduledFuture<?> heartbeatFuture;
    private final HashMap<Long, SimpleEntry<TurmsRequest, CompletableFuture<TurmsNotification>>> requestMap = new HashMap<>();

    private List<Function<TurmsNotification, Void>> onNotificationListeners = new LinkedList<>();
    private Function4<Boolean, TurmsCloseStatus, String, Throwable, Void> onClose;

    private String websocketUrl = "ws://localhost:9510";
    private int connectionTimeout = 10 * 1000;
    private int minRequestsInterval = 0;
    private long lastRequestDate = 0;
    private Long userId;
    private String password;
    private UserLocation userLocation;
    private UserStatus userOnlineStatus;
    private DeviceType deviceType;

    private String address;
    private String sessionId;

    public List<Function<TurmsNotification, Void>> getOnNotificationListeners() {
        return onNotificationListeners;
    }

    public String getAddress() {
        return address;
    }

    public String getSessionId() {
        return sessionId;
    }

    public void setOnClose(Function4<Boolean, TurmsCloseStatus, String, Throwable, Void> onClose) {
        this.onClose = onClose;
    }

    public TurmsDriver(@Nullable String websocketUrl,
                       @Nullable Integer connectionTimeout,
                       @Nullable Integer minRequestsInterval) {
        if (websocketUrl != null) {
            this.websocketUrl = websocketUrl;
        }
        if (connectionTimeout != null) {
            this.connectionTimeout = connectionTimeout;
        }
        if (minRequestsInterval != null) {
            this.minRequestsInterval = minRequestsInterval;
        }
        this.heartbeatInterval = HEARTBEAT_INTERVAL;
    }

    public CompletableFuture<Void> sendHeartbeat() {
        if (this.connected()) {
            lastRequestDate = System.currentTimeMillis();
            return websocket.sendBinary(ByteBuffer.allocate(0), true)
                    .thenApply(webSocket -> null);
        } else {
            return CompletableFuture.failedFuture(TurmsBusinessException.get(TurmsStatusCode.CLIENT_SESSION_HAS_BEEN_CLOSED));
        }
    }

    public boolean connected() {
        return websocket != null && !websocket.isInputClosed() && !websocket.isOutputClosed();
    }

    public CompletableFuture<Void> disconnect() {
        if (connected()) {
            return this.websocket
                    .sendClose(WebSocket.NORMAL_CLOSURE, "")
                    .thenApply(webSocket -> null);
        } else {
            return CompletableFuture.failedFuture(TurmsBusinessException.get(TurmsStatusCode.CLIENT_SESSION_HAS_BEEN_CLOSED));
        }
    }

    public CompletableFuture<Void> connect(
            long userId,
            @NotNull String password,
            @Nullable Integer connectionTimeout,
            @Nullable UserLocation userLocation,
            @Nullable UserStatus userOnlineStatus,
            @Nullable DeviceType deviceType) {
        if (connectionTimeout == null) {
            connectionTimeout = this.connectionTimeout;
        }
        Integer finalConnectionTimeout = connectionTimeout;
        if (connected()) {
            return CompletableFuture.failedFuture(TurmsBusinessException.get(TurmsStatusCode.CLIENT_SESSION_ALREADY_ESTABLISHED));
        } else {
            this.userId = userId;
            this.password = password;
            this.userLocation = userLocation;
            this.userOnlineStatus = userOnlineStatus;
            this.deviceType = deviceType;
            long connectionRequestId = (long) Math.ceil(Math.random() * Long.MAX_VALUE);
            WebSocket.Builder builder = HttpClient.newHttpClient()
                    .newWebSocketBuilder();
            builder.header(REQUEST_ID_FIELD, String.valueOf(connectionRequestId));
            builder.header(USER_ID_FIELD, String.valueOf(userId));
            builder.header(PASSWORD_FIELD, password);
            if (userLocation != null) {
                String location = String.format("%f%s%f", userLocation.getLongitude(), LOCATION_SPLIT, userLocation.getLatitude());
                builder.header(USER_LOCATION_FIELD, location);
            }
            if (userOnlineStatus != null) {
                builder.header(USER_ONLINE_STATUS_FIELD, userOnlineStatus.toString());
            }
            if (deviceType != null) {
                builder.header(DEVICE_TYPE_FIELD, deviceType.name());
            }
            return builder
                    .connectTimeout(Duration.ofSeconds(finalConnectionTimeout))
                    .buildAsync(URI.create(websocketUrl), new WebSocket.Listener() {

                        @Override
                        public void onOpen(WebSocket webSocket) {
                            webSocket.request(1);
                            onWebsocketOpen();
                        }

                        @Override
                        public CompletionStage<?> onBinary(WebSocket webSocket, ByteBuffer data, boolean last) {
                            webSocket.request(1);
                            TurmsNotification notification;
                            try {
                                notification = TurmsNotification.parseFrom(data);
                            } catch (InvalidProtocolBufferException e) {
                                TurmsLogger.logger.log(Level.SEVERE, "", e);
                                return CompletableFuture.failedStage(e);
                            }
                            if (notification != null) {
                                boolean isSessionInfo = notification.hasData() && notification.getData().hasSession();
                                if (isSessionInfo) {
                                    address = notification.getData().getSession().getAddress();
                                    sessionId = notification.getData().getSession().getSessionId();
                                } else if (notification.hasRequestId()) {
                                    long requestId = notification.getRequestId().getValue();
                                    SimpleEntry<TurmsRequest, CompletableFuture<TurmsNotification>> pair = requestMap.get(requestId);
                                    if (pair != null) {
                                        CompletableFuture<TurmsNotification> future = pair.getValue();
                                        if (notification.hasCode()) {
                                            int code = notification.getCode().getValue();
                                            if (TurmsStatusCode.isSuccess(code)) {
                                                future.complete(notification);
                                            } else {
                                                TurmsBusinessException exception;
                                                if (code == TurmsStatusCode.FAILED.getBusinessCode()) {
                                                    exception = TurmsBusinessException.get(code, notification.getReason().getValue());
                                                } else {
                                                    exception = TurmsBusinessException.get(code);
                                                }
                                                if (exception != null) {
                                                    future.completeExceptionally(exception);
                                                } else {
                                                    TurmsLogger.logger.log(Level.WARNING, "Unknown status code");
                                                }
                                            }
                                        } else {
                                            future.complete(notification);
                                        }
                                    }
                                }
                                for (Function<TurmsNotification, Void> listener : onNotificationListeners) {
                                    try {
                                        listener.apply(notification);
                                    } catch (Exception e) {
                                        e.printStackTrace();
                                    }
                                }
                            }
                            return CompletableFuture.completedStage(notification);
                        }

                        @Override
                        public CompletionStage<?> onClose(WebSocket webSocket, int statusCode, String reason) {
                            webSocket.request(1);
                            onWebsocketClose(statusCode, reason);
                            return null;
                        }

                        @Override
                        public void onError(WebSocket webSocket, Throwable error) {
                            onWebsocketError(error);
                        }
                    })
                    .whenComplete((webSocket, throwable) -> {
                        if (throwable != null) {
                            Throwable cause = throwable.getCause();
                            // Note that the WebSocketHandshakeException cannot be caught by onError
                            if (cause instanceof WebSocketHandshakeException) {
                                WebSocketHandshakeException handshakeException = (WebSocketHandshakeException) cause;
                                int statusCode = handshakeException.getResponse().statusCode();
                                if (statusCode == 307) {
                                    String address = handshakeException.getResponse()
                                            .headers().firstValue("reason").get();
                                    reconnect(address);
                                }
                            }
                        } else if (webSocket != null) {
                            this.websocket = webSocket;
                        }
                    })
                    .thenApply(webSocket -> null);
        }
    }

    public CompletableFuture<TurmsNotification> send(Message.Builder builder, Map<String, ?> fields) {
        if (fields != null) {
            ProtoUtil.fillFields(builder, fields);
        }
        Descriptors.Descriptor descriptor = builder.getDescriptorForType();
        String fieldName = StringUtil.camelToSnakeCase(descriptor.getName());
        TurmsRequest.Builder requestBuilder = TurmsRequest.newBuilder();
        Descriptors.Descriptor requestDescriptor = requestBuilder.getDescriptorForType();
        Descriptors.FieldDescriptor fieldDescriptor = requestDescriptor.findFieldByName(fieldName);
        requestBuilder.setField(fieldDescriptor, builder.build());
        return send(requestBuilder);
    }

    public CompletableFuture<TurmsNotification> send(TurmsRequest.Builder requestBuilder) {
        if (this.connected()) {
            Date now = new Date();
            if (minRequestsInterval == 0 || now.getTime() - lastRequestDate > minRequestsInterval) {
                lastRequestDate = now.getTime();
                long requestId = generateRandomId();
                requestBuilder.setRequestId(Int64Value.newBuilder().setValue(requestId).build());
                TurmsRequest request = requestBuilder.build();
                ByteBuffer data = ByteBuffer.wrap(request.toByteArray());
                CompletableFuture<TurmsNotification> future = new CompletableFuture<>();
                requestMap.put(requestId, new SimpleEntry<>(request, future));
                websocket.sendBinary(data, true).whenComplete((webSocket, throwable) -> {
                    if (throwable != null) {
                        future.completeExceptionally(throwable);
                    }
                });
                return future;
            } else {
                return CompletableFuture.failedFuture(TurmsBusinessException.get(TurmsStatusCode.CLIENT_REQUESTS_TOO_FREQUENT));
            }
        } else {
            return CompletableFuture.failedFuture(TurmsBusinessException.get(TurmsStatusCode.CLIENT_SESSION_HAS_BEEN_CLOSED));
        }
    }

    private long generateRandomId() {
        long id;
        do {
            id = (long) Math.floor(Math.random() * Long.MAX_VALUE);
        } while (requestMap.containsKey(id));
        return id;
    }

    private void onWebsocketOpen() {
        isSessionEstablished = true;
        heartbeatFuture = this.heartbeatTimer.scheduleAtFixedRate(
                this::checkAndSendHeartbeatTask,
                heartbeatInterval,
                heartbeatInterval,
                TimeUnit.SECONDS);
    }

    private void checkAndSendHeartbeatTask() {
        long difference = System.currentTimeMillis() - lastRequestDate;
        if (difference > minRequestsInterval) {
            this.sendHeartbeat();
        }
    }

    private void onWebsocketClose(int code, String reason) {
        boolean wasLogged = isSessionEstablished;
        isSessionEstablished = false;
        cancelHeartbeatFuture();
        TurmsCloseStatus status = TurmsCloseStatus.get(code);
        if (status == TurmsCloseStatus.REDIRECT) {
            reconnect(reason);
        } else if (onClose != null) {
            if (status == TurmsCloseStatus.WEBSOCKET_ERROR) {
                reason = String.valueOf(code);
            }
            onClose.apply(wasLogged, status, reason, null);
        }
    }

    private void onWebsocketError(Throwable error) {
        boolean wasLogged = isSessionEstablished;
        isSessionEstablished = false;
        cancelHeartbeatFuture();
        onClose.apply(wasLogged, null, null, error);
    }

    private void reconnect(String address) {
        boolean isSecure = websocketUrl.startsWith("wss://");
        websocketUrl = (isSecure ? "wss://" : "ws://") + address;
        this.connect(userId, password, connectionTimeout, userLocation, userOnlineStatus, deviceType);
    }

    private void cancelHeartbeatFuture() {
        if (!heartbeatFuture.isCancelled() && !heartbeatFuture.isDone()) {
            heartbeatFuture.cancel(true);
        }
    }
}
